src/async_fuse/memfs/kv_engine/etcd_impl.rs

```diff
@@ -292,6 +616,102 @@ impl MetaTxn for EtcdTxn {
         self.buffer.insert(key, None);
     }
 
+    /// Try to campaign the master with simple txn
+    /// Old(reference to etcd client v3):
+    /// Create a key with prefix, and compare key (sorted by revision) with the key created by the same session.
+    /// If current key is the smallest, then the session is the master, and campaign success.
+    ///
+    /// New(For datenlord cache scenario):
+    /// A small batch of nodes in cluster, we just need a simple txn to get the master key.
+    /// The thundering herd is not the main problem, which will cause a bunch of etcd raft logs.
+    /// Use String as val can be easily compared.
+    async fn campaign(
+        &self,
+        key: &KeyType,
+        val: String,
+        lease_id: i64,
+    ) -> DatenLordResult<(bool, Option<String>)> {
+        let mut client = self.client.clone();
+
+        // Try to get the key, if key is existed
+        // We need to return the data from etcd,
+        // If the key is not existed, we need to set the key.
+        let txn = Txn::new()
+            // Check the key is existed and key is not equal to current value, which means the key is ready to be campaign.
+            .when(vec![
+                Compare::create_revision(key.to_string_key(), CompareOp::NotEqual, 0),
+                Compare::value(key.to_string_key(), CompareOp::NotEqual, val.clone()),
+            ])
+            // Campaign failed
+            .and_then(vec![TxnOp::get(
+                key.to_string_key(),
+                Some(GetOptions::new().with_serializable()),
+            )])
+            // Campaign succeed
+            .or_else(vec![TxnOp::put(
+                key.to_string_key(),
+                val.clone(),
+                Some(PutOptions::new().with_lease(lease_id)),
+            )]);
+
+        let (campaign_status, responses) = match client.txn(txn).await {
+            Ok(resp) => {
+                // Check the txn branch is `then` or `else`
+                // If succeeded, means the campaign is failed and can not set current data to etcd
+                // If failed, means the campaign is success, and put the current data to etcd
+                (!resp.succeeded(), resp.op_responses())
+            }
+            Err(e) => {
+                error!("failed to campaign, error={e:?}");
```

Comment:
Here encounter etcd error,  return Err() directly, no need to fall through,

Reply:
Ok, I have updated it to err.